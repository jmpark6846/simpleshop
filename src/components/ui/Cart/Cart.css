.cart{ 
  position: absolute;
  background-color: white;
  top: 53px;
  right: 15px;
  border:1px solid lightgray;
  border-radius: 3px;
  display: none;
  font-size: 0.9rem;
  width: 300px;
  z-index: 1000;
}


.cart.show{
  display: block;
}

.cart-item, .cart-item-wrapper{
  padding: 15px;
  
  border-bottom: 1px solid lightgray;
  display: flex;
}
.cart-item .image{
  background-size: cover;
  background-repeat: no-repeat;
  background-position: center;
  width:30%;
  display: inline-block;
}

.cart-item .image::after{
  /* .slider .slide::after{ */
    content:'';
    display: block;
    z-index: -100;
    padding-bottom: 100%;
  }

.cart-item .item-info{
  width:70%;
  display: inline-block;
  font-size: 0.9rem;
  vertical-align: top;
  padding: 0 5px;
}

.cart .error{
  background-color: rgba(216, 112, 147);
  color:white;
}

.cart .icon{
  margin-right: 5px;
}
.cart .delete{
  align-self: center;
}
.cart .delete .icon{
  cursor: pointer;
}

.cart button.buy{
  width:100%;
}