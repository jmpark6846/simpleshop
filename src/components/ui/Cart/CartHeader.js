import React from 'react'
import PropTypes from 'prop-types'

const CartHeader = ({value}) => {
  return (
    <div className='cart-header'>
      {value}
    </div>
  )
}

CartHeader.propTypes = {
  value: PropTypes.string
}

export default CartHeader
