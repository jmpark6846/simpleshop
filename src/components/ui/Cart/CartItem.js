import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { NavLink } from "react-router-dom";
import NumberSelector from '../../ui-components/NumberSelector/NumberSelector'
import Price from "../../ui-components/Price/Price";
import Icon from '../../ui-components/Icon/Icon'
import { doDeleteCartItem, doChangeEA } from '../../../actions/cart';

export const CartItem = ({cartItem, deleteCartItem, changeEA}) => {
  return (
    <div className='cart-item'>
      <div className='image' style={{ backgroundImage: `url(${cartItem.img})` }}></div>      
      <div className='item-info'>
        <div><NavLink to={'/product/'+cartItem.id} >{ cartItem.name.length > 20 ? `${cartItem.name.substr(0,20)}...` : cartItem.name }</NavLink></div>
        <div><Price price={ cartItem.price } suffix='원' /></div>
        <NumberSelector number={cartItem.ea} onChange={(ea)=>changeEA(cartItem.id, ea)} />
      </div>
      <div className='delete'>
        <Icon icon='far fa-trash-alt' onClick={()=>deleteCartItem(cartItem.id)} />
      </div>
    </div>
  )
}

CartItem.propTypes = {
  className: PropTypes.string,
  children: PropTypes.node,
  changeEA: PropTypes.func,
  deleteCartItem: PropTypes.func,
  cartItem: PropTypes.object
}

const mapDispatch = (dispatch) => ({
  changeEA: (id, ea) => dispatch(doChangeEA(id, ea)),
  deleteCartItem: (id) => dispatch(doDeleteCartItem(id))
})
export default connect(undefined, mapDispatch)(CartItem)
