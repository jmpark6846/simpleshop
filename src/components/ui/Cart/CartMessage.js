import React from 'react'
import PropTypes from 'prop-types'
import Icon from '../../ui-components/Icon/Icon';

class CartMessage extends React.Component{
  render(){
    const { children, show, handleCloseClick } = this.props
    return (
      <div className={ show ? 'cart-message' : 'cart-message hide'} >
        <div className='cart-message-wrapper'>
          <div className='cart-message-body'>
            { children ? children : '장바구니에 추가되었습니다.'}
            <span onClick={handleCloseClick} className='icon'>×</span>
          </div>
          <i className='chat-start-icon'></i>
        </div>
      </div>
    )
  }
  
}

CartMessage.propTypes = {
  children: PropTypes.node,
  show: PropTypes.bool,
  handleCloseClick: PropTypes.func
}

export default CartMessage
