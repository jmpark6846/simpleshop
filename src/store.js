import { createStore, applyMiddleware, combineReducers } from 'redux'
import { createLogger } from 'redux-logger'
import { cart, product, order } from './reducers'

const root = combineReducers({
  cart: cart,
  product: product,
  order: order,
})

let store=null

if(process.env.NODE_ENV === 'development'){
  store = createStore(root, undefined, applyMiddleware(createLogger()))  
}else{  
  store = createStore(root, undefined)  
}

export default store