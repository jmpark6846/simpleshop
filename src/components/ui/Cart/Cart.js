import React from 'react'
import PropTypes from 'prop-types'
import { connect } from "react-redux";
import classNames from 'classnames'
import './Cart.css'
import CartItem from './CartItem';

import Icon from '../../ui-components/Icon/Icon';
import Button from '../../ui-components/Button/Button';
import Price from '../../ui-components/Price/Price';

export const Cart = ({show, cartItems, totalPrice}) => {
  return (
    <div className='cart-wrapper'>
      <Button className='shopping-cart' icon='fas fa-shopping-cart'/>
      <div className='cart-preview-div'>
        <div className={classNames('cart', { show })} >
          { cartItems &&Object.keys(cartItems).length !== 0 ? 
            Object.keys(cartItems).map(id=> <CartItem key={id} cartItem={cartItems[id]} />)          
            :
            <div className='cart-item-wrapper cart-empty-message'>장바구니가 비었습니다.</div> 
          }
        
          <div className='cart-item-wrapper'>
            총 금액: <Price className='cart-total-price' price={totalPrice} suffix='원'/>
          </div> 
          <div className='cart-item-wrapper'>
            <Button className='buy' value='구매하기' primary/>
          </div>
        </div>
      </div>  
    </div>
  )
}

Cart.propTypes = {
  show: PropTypes.bool,
  cartItems: PropTypes.object,
  totalPrice: PropTypes.number
}

const mapState = ({cart}) => ({
  cartItems: cart.cartItems,
  totalPrice: cart.totalPrice
})

export default connect(mapState)(Cart)
