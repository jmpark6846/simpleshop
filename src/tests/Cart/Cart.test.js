import React from 'react'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { Cart } from '../../components/ui/Cart/Cart';
import ConnectedCartItem from '../../components/ui/Cart/CartItem';
import store from "../../store";

Enzyme.configure({ adapter: new Adapter()})

describe('Cart', ()=> {
  const cartItems = { 
    0 : { id:0, name: 'bag1', price: 500, ea:1, img:'/src/images/bag1.jpg', },
    1 : { id:1, name: 'bag2', price: 400, ea:2, img:'/src/images/bag2.jpg', },
    2 : { id:2, name: 'bag3', price: 600, ea:3, img:'/src/images/bag3.jpg', },
  }
  
  it('Cart snapshot test', () =>{
    const tree = renderer.create(<Provider store={store}><Cart cartItems={cartItems} /></Provider>).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('카트 항목에 맞게 카트 리스트를 렌더링', () => {  
    const wrapper = shallow(<Cart cartItems={cartItems} />)
    expect(wrapper.find(ConnectedCartItem).length).toBe(5) // 카트 아이템 3개 + 총금액 div 1개 + 구매 버튼 div 1개
  })
})