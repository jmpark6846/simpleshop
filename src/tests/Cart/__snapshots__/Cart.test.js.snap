// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Cart snapshot test 1`] = `
<div
  className="cart"
>
  <div
    className="cart-item"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(/src/images/bag1.jpg)",
        }
      }
    />
    <div
      className="item-info"
    >
      <div>
        bag1
      </div>
      <div>
        <span>
          500
          원
        </span>
      </div>
      <div
        className="number-selector"
      >
        <button
          className="decrement"
          onClick={[Function]}
        >
          -
        </button>
        <input
          className="number"
          name="number"
          onChange={[Function]}
          value={1}
        />
        <button
          className="increment"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="delete"
    >
      <span
        className="icon far fa-trash-alt"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="cart-item"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(/src/images/bag2.jpg)",
        }
      }
    />
    <div
      className="item-info"
    >
      <div>
        bag2
      </div>
      <div>
        <span>
          400
          원
        </span>
      </div>
      <div
        className="number-selector"
      >
        <button
          className="decrement"
          onClick={[Function]}
        >
          -
        </button>
        <input
          className="number"
          name="number"
          onChange={[Function]}
          value={2}
        />
        <button
          className="increment"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="delete"
    >
      <span
        className="icon far fa-trash-alt"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="cart-item"
  >
    <div
      className="image"
      style={
        Object {
          "backgroundImage": "url(/src/images/bag3.jpg)",
        }
      }
    />
    <div
      className="item-info"
    >
      <div>
        bag3
      </div>
      <div>
        <span>
          600
          원
        </span>
      </div>
      <div
        className="number-selector"
      >
        <button
          className="decrement"
          onClick={[Function]}
        >
          -
        </button>
        <input
          className="number"
          name="number"
          onChange={[Function]}
          value={3}
        />
        <button
          className="increment"
          onClick={[Function]}
        >
          +
        </button>
      </div>
    </div>
    <div
      className="delete"
    >
      <span
        className="icon far fa-trash-alt"
        onClick={[Function]}
      />
    </div>
  </div>
  <div
    className="cart-item"
  >
    총 금액: 
    <span
      className="cart-total-price"
    >
      und,efi,ned
      원
    </span>
  </div>
  <div
    className="cart-item"
  >
    <button
      className="button buy primary"
    >
      구매하기
    </button>
  </div>
</div>
`;
